File: backend/src/029-concurrencia/hilos/join/_join.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import time
import threading

# =====================================

def contador():
    for i in range(1, 6):
        print(i)
        time.sleep(1)

# Crear hilos
t1 = threading.Thread(target=contador)
t2 = threading.Thread(target=contador)

# Iniciar la ejecución
t1.start()
t2.start()

# Esperar a que se ejecuten los hilos
t1.join()
t2.join()

print("Fin de la ejecución")

# =====================================

# Obtener el hilo actual
actual = threading.current_thread()

# Obtener el hilo principal
principal = threading.main_thread()

# Lista con los hilos activos
hilos = threading.enumerate()

# Número de hilos activos
threading.active_count() # 1

# Identificador del hilo actual
threading.get_ident()

# ====================================

# Fuente:
# https://docs.python.org/3/library/threading.html

File: backend/src/029-concurrencia/hilos/join/_join.cpp
#include <iostream>
#include <thread>
#include <chrono>

// Para compilar, linkear con -pthread:
// g++ th.cpp -o t -pthread

using namespace std;

void dormir(int segundos) {
    this_thread::sleep_for(chrono::seconds(segundos));
}

// Función que se ejecuta al crear el hilo
void contador() {
    for (int i=1; i<6; ++i) {
        cout << i << "\n";
        dormir(1);
    }
}

int main() {
    // Crear hilos
    thread t1{contador};
    thread t2(contador);

    // Esperamos a que terminen los hilos
    t1.join();
    t2.join();

    cout << "Fin de la ejecución" << "\n";

    return 0;
}

/* Fuentes:
http://www.cplusplus.com/reference/thread/thread/
http://www.cplusplus.com/reference/thread/this_thread/sleep_for/
https://solarianprogrammer.com/2011/12/16/cpp-11-thread-tutorial/
*/